### DISCUSSION TITLE
Sort imports below `__all__`

### DISCUSSION THREAD

**Discussion by JacobCoffee** [astral-sh/ruff#8540] [Ideas]
When using context UI importing in PyCharm sometimes imports get place below `__all__`.

I'm just wondering, even if its not "expected behavior for iSort" or whatever, if it's possible to sort this block in the event imports are below `__all__`.

If `if TYPE_CHECKING:` is below `__all__`, we get the same behavior - but if it is above then things import correctly.

Maybe we could sort the blocks so that we have

1. Imports
2. `if TYPE_CHECKING:`
3. `__all__`
4. bad code
?

<details>
<summary>Before:</summary>

```py
from __future__ import annotations

import click
from rich import get_console

from src.lib import _quote

__all__ = [
    "some_func",
    "run",
]
```
</details>

<details>
<summary>Using the little context UI to import:</summary>

<img width="769" alt="image" src="https://github.com/astral-sh/ruff/assets/45884264/b3589677-7c11-4f26-89b2-4d4d11160110">

</details>

<details>
<summary>The now imported and Ruff sorted block:</summary>

```py
from __future__ import annotations

import click
from rich import get_console

from src.lib import _quote

__all__ = [
    "some_func",
    "run",
]

from src.lib import logging

logger = logging.getLogger(__name__)
```

</details>

However, when I am using `if TYPE_CHECKING:` (above `__all__` only, below is the same behavior) it correctly imports.

<details>
<summary>Doing the same thing but with `if TYPE_CHECKING:`:</summary>

```py
from __future__ import annotations

import click
from rich import get_console

from src.lib import _quote
from src.lib import logging

if TYPE_CHECKING:
    from java import bad

__all__ = [
    "some_func",
    "run",
]

logger = logging.getLogger(__name__)
```
</details>

**Comment by zanieb** [comment]
Interesting. I'm not sure if we can reliably sort imports that are in separate blocks. This sounds like a PyCharm bug though?

**Reply by JacobCoffee** [reply]
> I'm not sure if we can reliably sort imports that are in separate blocks.
darn 😬 

Yeah I think so... it surely isn't an issue in the correct editor (neovim), or VSCode 🙁

**Comment by JacobCoffee** [comment]
https://youtrack.jetbrains.com/issue/PY-63973 this was reproduced on Jetbrains' side.
